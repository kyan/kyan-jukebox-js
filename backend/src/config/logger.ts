import { Logger } from 'tslog'

const log = new Logger({
  type: process.env.NODE_ENV === 'production' ? 'json' : 'pretty',
  minLevel: 0,
  hideLogPositionForProduction: process.env.NODE_ENV === 'production'
})

export default log
