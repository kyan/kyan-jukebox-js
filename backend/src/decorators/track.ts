import { JBTrack } from '../types/database'

export default function (json: any): JBTrack {
  if (!json) throw new Error('DecorateTrack passed no data!')

  const payload: JBTrack = {
    uri: json.uri,
    name: json.name,
    length: json.length || json.duration_ms
  }

  if (json.image) {
    payload.image = json.image
  }

  if (json.album) {
    payload.album = {
      uri: json.album.uri,
      name: json.album.name,
      year: json.album.date
    }

    if (!payload.image && json.album.images && json.album.images.length > 0) {
      payload.image = json.album.images[0].url
    }
  }

  payload.artist = {
    uri: json.artists[0].uri,
    name: json.artists[0].name
  }

  if (json.date) {
    payload.year = json.date
  }

  if (json.addedBy) {
    payload.addedBy = json.addedBy
  }

  if (json.metrics) {
    payload.metrics = json.metrics
  }

  if (process.env.EXPLICIT_CONTENT === 'false' && json.explicit) {
    payload.explicit = true
  }

  return payload
}
