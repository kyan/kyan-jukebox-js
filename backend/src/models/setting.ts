import { Schema, model } from 'mongoose'

const SettingSchema = new Schema({
  key: Schema.Types.String,
  value: Schema.Types.Mixed
})

const Setting = model('Setting', SettingSchema)

export default Setting
