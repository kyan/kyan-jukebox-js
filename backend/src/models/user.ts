import { Document, Schema, model } from 'mongoose'

export interface JBUser {
  _id: any
  fullname: string
  picture: string
}

interface User extends JBUser, Document {
  _id: any
}

const userSchema = new Schema(
  {
    _id: Schema.Types.String,
    fullname: Schema.Types.String,
    picture: Schema.Types.String
  },
  { _id: false }
)
const User = model<User>('User', userSchema)

export default User
