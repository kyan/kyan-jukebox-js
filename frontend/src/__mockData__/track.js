const MockTrackJson = () => {
  return {
    uri: 'spotify:track:1WaBuuaXGwrU4sFvxAjnkf',
    name: 'Happiness',
    length: 120000,
    album: {
      uri: 'spotify:album:1D7E2EHMoQM3nP9drEwv7o',
      name: 'Ken Dodd',
      year: '1991'
    },
    artist: {
      uri: 'spotify:artist:76o4kCpWMmBGl8jIYfRHTk',
      name: 'Ken Dodd'
    },
    year: '1991',
    explicit: true,
    image: 'https://i.scdn.co/image/mock-track-image.jpg'
  }
}

export default MockTrackJson
