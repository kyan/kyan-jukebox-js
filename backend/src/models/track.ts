import Mopidy from 'mopidy'
import { Schema, model, Model, Document } from 'mongoose'
import User, { JBUser } from '../models/user'
import EventLogger from '../utils/event-logger'
import logger from '../config/logger'
import VotingHelper from '../utils/voting'

export interface JBVotes {
  user: JBUser
  vote: number
  at: Date
}

interface JBPlayed {
  at: Date
}

export interface JBAddedBy {
  user: JBUser
  addedAt: Date
  played: JBPlayed[]
  votes: JBVotes[]
}

interface JBMetrics {
  plays: number
  votes: number
  votesTotal: number
  votesAverage: number
}

interface Track extends Document {
  _id: string
  uri: string
  addedBy: JBAddedBy[]
  metrics: JBMetrics
}

const TrackSchema = new Schema(
  {
    _id: {
      type: Schema.Types.String,
      alias: 'uri'
    },
    addedBy: [
      {
        _id: false,
        user: { type: Schema.Types.String, ref: 'User' },
        addedAt: Schema.Types.Date,
        played: [
          {
            _id: false,
            at: Schema.Types.Date
          }
        ],
        votes: [
          {
            _id: false,
            user: { type: Schema.Types.String, ref: 'User' },
            vote: Schema.Types.Number,
            at: Schema.Types.Date
          }
        ]
      }
    ],
    metrics: {
      _id: false,
      plays: { type: Schema.Types.Number, default: 0 },
      votes: { type: Schema.Types.Number, default: 0 },
      votesTotal: { type: Schema.Types.Number, default: 0 },
      votesAverage: { type: Schema.Types.Number, default: 0 }
    }
  },
  { _id: false }
)

export interface MopidyImageExt extends Mopidy.models.Image {
  url: string
}

export interface MopidyAlbumExt extends Mopidy.models.Album {
  images: MopidyImageExt[]
}

export interface MopidyTrackExt extends Mopidy.models.Track {
  album: MopidyAlbumExt
  addedBy?: JBAddedBy[]
  metrics?: JBMetrics
  image?: string
  genres?: string[]
  duration_ms?: number
  explicit?: boolean
}

export interface SpotifyTrackObjectFullExt extends SpotifyApi.TrackObjectFull {
  addedBy: JBAddedBy[]
  metrics: JBMetrics
}

export interface JBWTrack {
  track: JBTrack
}

export interface JBArtist {
  uri: string
  name: string
}

export interface JBAlbum {
  uri: string
  name: string
  year: string
}

export interface JBTrack {
  uri: string
  name: string
  length: number
  year?: string
  image?: string
  album?: JBAlbum
  artist?: JBArtist
  addedBy?: any
  metrics?: any
  explicit?: boolean
  genres?: string[]
}

const brh = {
  _id: '1ambigrainbowhead',
  fullname: 'BRH',
  picture: 'https://cdn-images-1.medium.com/fit/c/200/200/1*bFBXYvskkPFI9nPx6Elwxg.png'
}

TrackSchema.statics.findTracks = (uris: string[]): Promise<Track[]> =>
  new Promise((resolve, reject) => {
    Track.find({ _id: { $in: uris } })
      .populate({ path: 'addedBy.user' })
      .populate({ path: 'addedBy.votes.user' })
      .then((tracks) => {
        if (tracks.length > 0) EventLogger.info('FOUND CACHED TRACKS', { data: uris })
        resolve(tracks)
      })
      .catch((err) => reject(err))
  })

TrackSchema.statics.findOrUseBRH = (user?: JBUser): Promise<any> => {
  if (user) return Promise.resolve(new User(user).toObject())

  return User.findOneAndUpdate({ _id: brh._id }, brh, {
    upsert: true,
    new: true,
    setDefaultsOnInsert: true
  }).exec()
}

TrackSchema.statics.addTracks = (
  uris: ReadonlyArray<string>,
  user?: JBUser
): Promise<any> =>
  new Promise((resolve) => {
    Track.findOrUseBRH(user)
      .then((returnUser: any) => {
        const requests = uris.map((uri) =>
          Track.findOneAndUpdate(
            { _id: uri },
            {
              $push: {
                addedBy: {
                  $each: [
                    {
                      user: returnUser,
                      addedAt: new Date(),
                      played: [],
                      votes: []
                    }
                  ],
                  $position: 0
                }
              }
            },
            { upsert: true, runValidators: true, setDefaultsOnInsert: true }
          ).exec()
        )

        Promise.all(requests)
          .then(() => resolve({ uris, user: returnUser }))
          .catch((error) =>
            logger.error('addTracks:Track.updateOne', { message: error.message })
          )
      })
      .catch((error: any) =>
        logger.error('addTracks:findOrUseBRH', { message: error.message })
      )
  })

const updateTrackPlaycount = (uri: string): Promise<Track> =>
  new Promise((resolve) => {
    Track.findById(uri)
      .populate({ path: 'addedBy.user' })
      .populate({ path: 'addedBy.votes.user' })
      .then((track) => {
        if (track && track.addedBy[0]) {
          const played: JBPlayed = { at: new Date() }
          track.addedBy[0].played.unshift(played)
          track.metrics.plays = track.metrics.plays + 1
          return track.save()
        }

        return Promise.resolve(track)
      })
      .then((track) => resolve(track))
      .catch((error) => logger.error('updateTrackPlaycount', { message: error.message }))
  })

const updateTrackVote = (uri: string, user: JBUser, vote: number) =>
  new Promise((resolve) => {
    Track.findOrUseBRH(user)
      .then((vUser: any) => {
        Track.findById(uri)
          .populate({ path: 'addedBy.user' })
          .populate({ path: 'addedBy.votes.user' })
          .then((track) => {
            if (track && track.addedBy[0]) {
              const currentVote = VotingHelper.voteNormalised(vote)
              const newVote: JBVotes = {
                at: new Date(),
                vote: currentVote,
                user: vUser
              }
              const votes = track.addedBy[0].votes
              const currentVoteIndex = votes.findIndex(
                (vote) => vote.user._id === vUser._id
              )

              if (currentVoteIndex !== -1) {
                votes[currentVoteIndex].vote = currentVote
              } else {
                votes.unshift(newVote)
                track.metrics.votes = VotingHelper.calcVoteCount(track.addedBy)
              }

              track.metrics.votesTotal = VotingHelper.calcVoteTotal(track.addedBy)
              track.metrics.votesAverage = VotingHelper.calcWeightedMean(track.addedBy)

              return track.save()
            }

            return Promise.resolve(track)
          })
          .then((track) =>
            resolve({ uri: track.id, addedBy: track.addedBy, metrics: track.metrics })
          )
          .catch((error) =>
            logger.error('updateTrackVote:findById', { message: error.message })
          )
      })
      .catch((error: { message: string }) =>
        logger.error('updateTrackVote:findOrUseBRH', { message: error.message })
      )
  })

interface DBTrackStatics extends Model<Track> {
  /**
   * Lookup tracks in MongoDB
   *
   * @param uris - A list of Track uris
   */
  findTracks(uris: ReadonlyArray<string>): Promise<Track[]>
  findOrUseBRH(user?: JBUser): Promise<any>
  addTracks(uris: ReadonlyArray<string>, user?: JBUser): Promise<any>
}

const Track = model<Track, DBTrackStatics>('Track', TrackSchema)

export default Track
export { updateTrackPlaycount, updateTrackVote }
