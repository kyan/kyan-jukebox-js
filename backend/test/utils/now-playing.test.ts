import Setting from '../../src/models/setting'
import logger from '../../src/config/logger'
import NowPlaying from '../../src/utils/now-playing'
import { JBTrack } from '../../src/models/track'
jest.mock('../../src/config/logger')

describe('NowPlaying', () => {
  describe('addTrack', () => {
    const trackObject: JBTrack = {
      uri: 'uri000',
      name: 'Seasons (Waiting On You)',
      year: '1983',
      image: 'the-album-art.jpg',
      length: 1234,
      artist: {
        uri: 'uri123',
        name: 'Future Islands'
      },
      album: {
        uri: 'uri321',
        name: 'Singles',
        year: '2019'
      },
      metrics: {
        votesAverage: 80,
        votes: 2,
        plays: 2
      },
      addedBy: [
        {
          user: {
            fullname: 'Duncan'
          },
          addedAt: new Date(1582010703141)
        },
        {
          user: {
            fullname: 'BRH'
          },
          addedAt: new Date(1582000703141)
        }
      ]
    }

    it('returns the correct payload when full data', () => {
      const result = {} as any
      jest.spyOn(Setting, 'findOneAndUpdate').mockResolvedValue(result)
      jest.spyOn(global.Date, 'now').mockImplementation(() => 1582020703141)

      return NowPlaying.addTrack(trackObject).then((payload) => {
        expect(payload).toMatchSnapshot()
      })
    })

    it('returns the correct payload when full data 1', () => {
      const result = {} as any
      jest.spyOn(Setting, 'findOneAndUpdate').mockResolvedValue(result)
      jest.spyOn(global.Date, 'now').mockImplementation(() => 1582020703141)
      trackObject.metrics.plays = 1
      delete trackObject.addedBy[1]

      return NowPlaying.addTrack(trackObject).then((payload) => {
        expect(payload).toMatchSnapshot()
      })
    })

    it('handles errors', () => {
      jest.spyOn(Setting, 'findOneAndUpdate').mockRejectedValue(new Error('oooops'))
      NowPlaying.addTrack(trackObject)

      return new Promise<void>((resolve) => {
        setTimeout(() => {
          expect(logger.error).toHaveBeenCalledWith('NowPlaying.addTrack: oooops')
          resolve()
        }, 0)
      })
    })
  })
})
