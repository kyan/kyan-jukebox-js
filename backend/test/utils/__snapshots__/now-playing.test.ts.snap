// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NowPlaying addTrack returns the correct payload when full data 1`] = `
Object {
  "added_at": "3 hours ago",
  "added_by": "Duncan",
  "album": "Singles",
  "artist": "Future Islands",
  "image": "the-album-art.jpg",
  "last_played": "6 hours ago",
  "metrics": Object {
    "plays": 2,
    "rating": 3,
    "votes": 2,
  },
  "spotify": "https://open.spotify.com/track/uri000",
  "title": "Seasons (Waiting On You)",
  "year": "1983",
}
`;

exports[`NowPlaying addTrack returns the correct payload when partial full data 1`] = `
Object {
  "added_at": "3 hours ago",
  "added_by": "Duncan",
  "album": "Singles",
  "artist": "Future Islands",
  "image": "the-album-art.jpg",
  "last_played": null,
  "metrics": Object {
    "plays": 1,
    "rating": 3,
    "votes": 2,
  },
  "spotify": "https://open.spotify.com/track/uri000",
  "title": "Seasons (Waiting On You)",
  "year": "1983",
}
`;
